File: 0054-spiral-matrix/0054-spiral-matrix.py
class Solution(object):
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        result = []
        up = 0
        down = len(matrix) - 1
        left = 0
        right = len(matrix[0]) - 1
        while up <= down:
            for row in range(left, right + 1):
                result.append(matrix[up][row])
            up += 1
            if left <= right:
                for col in range(up, down + 1):
                    result.append(matrix[col][right])
                right -= 1
            if up <= down:
                for row in range(right, left -1, -1):
                    result.append(matrix[down][row])
                down -= 1
                if left <= right:
                    for col in range(down, up -1, -1):
                        result.append(matrix[col][left])
                    left += 1
        return result

File: 0054-spiral-matrix/0054-spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        int x = 0;
        int y = 0;
        int maxX = matrix[0].length-1;
        int maxY = matrix.length-1;
        
        List<Integer> answer = new ArrayList<>();
        
        while(x <= maxX && y <= maxY){
            for(int i = x; i <= maxX; i++){
                answer.add(matrix[y][i]);
            }
            y += 1;
            for(int j = y; j <= maxY; j++){
                answer.add(matrix[j][maxX]);
            }
            maxX -= 1;
            if(y <= maxY){
                for(int i = maxX; i >= x; i--){
                    answer.add(matrix[maxY][i]);
                }
                maxY -= 1;
            }
            if(x <= maxX){
                for(int j = maxY; j >= y; j--){
                     answer.add(matrix[j][x]);
                }
            x += 1;
            }  
            
        }
        
        return answer;
        
    }
}